File: credit.c
#include <cs50.h>
#include <stdio.h>

void determine_card_type(long cardNumber);
bool test_card_number(long cardNumber);
long Pow(long num, long power);

int main(void)
{
    bool exit = true;
    long cardNumber;
    
    while (exit)
    {
        cardNumber = get_long("Number: \n");
        if (test_card_number(cardNumber))
        {
            determine_card_type(cardNumber);
            exit = false;
        }
        else
        {
            exit = false;
        }
    }
}

// Determine card type.
void determine_card_type(long cardNumber)
{
    if ((cardNumber % Pow(10, 15) - cardNumber % Pow(10, 13)) / Pow(10, 13) == 34 ||
        (cardNumber % Pow(10, 15) - cardNumber % Pow(10, 13)) / Pow(10, 13) == 37)
    {
        printf("AMEX\n");
    }
    else if ((cardNumber - cardNumber % Pow(10, 14)) / Pow(10, 14) == 51 ||
             (cardNumber - cardNumber % Pow(10, 14)) / Pow(10, 14) == 52 ||
             (cardNumber - cardNumber % Pow(10, 14)) / Pow(10, 14) == 53 ||
             (cardNumber - cardNumber % Pow(10, 14)) / Pow(10, 14) == 54 ||
             (cardNumber - cardNumber % Pow(10, 14)) / Pow(10, 14) == 55)
    {
        printf("MASTERCARD\n");
    }
    else if ((cardNumber - cardNumber % Pow(10, 15)) / Pow(10, 15) == 4 ||
             (cardNumber % Pow(10, 14) - cardNumber % Pow(10, 13)) / Pow(10, 13) == 4)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

// Card number validation test.
bool test_card_number(long cardNumber)
{
    if (cardNumber < Pow(10, 17) && cardNumber >= Pow(10, 14))
    {
        long evenDigitsSum = 0;
        long evenDigits = 0;
        long oddDigits = 0;
        long oddDigitsSum = 0;
        for (long i = 100; i < Pow(10, 17); i *= 100)
        {
            evenDigits = (((cardNumber % i) - (cardNumber % (i / 10))) / (i / 10)) * 2;
            evenDigitsSum = evenDigitsSum + (evenDigits % 10 + ((evenDigits % 100) - (evenDigits % 10)) / 10);
            
            oddDigits = ((cardNumber % (i / 10)) - (cardNumber % (i / 100))) / (i / 100);
            oddDigitsSum = oddDigitsSum + oddDigits;
        } 
        
        if ((evenDigitsSum + oddDigitsSum) % 10 == 0)
        {
            return true;
        }
        else
        {
            return false;
        }
    }
    else
    {
        printf("INVALID\n");
        return false;
    }    
}

// Returns a specified number raised to the specified power.
long Pow(long num, long power)
{
    long result = num;
    for (int i = 0; i < power - 1; i++)
    {
        result = num * result;
    }
    return result;
}

File: credit.py
def main():
    exit = True
    card_num = 0

    while exit:
        print("Number: ", end="")
        card_num = int(input())
        if test_card_number(card_num):
            determine_card_type(card_num)
            exit = False
        else:
            print("INVALID")
            exit = False

# Determine card type.


def determine_card_type(card_num):
    if (card_num % 10**15 - card_num % 10**13) / 10**13 == 34 or (card_num % 10**15 - card_num % 10**13) / 10**13 == 37:
        print("AMEX")
    elif (card_num - card_num % 10**14) / 10**14 == 51 or (card_num - card_num % 10**14) / 10**14 == 52 or (card_num - card_num % 10**14) / 10**14 == 53 or (card_num - card_num % 10**14) / 10**14 == 54 or (card_num - card_num % 10**14) / 10**14 == 55:
        print("MASTERCARD")
    elif (card_num - card_num % 10**15) / 10**15 == 4 or (card_num % 10**14 - card_num % 10**13) / 10**13 == 4:
        print("VISA")
    else:
        print("INVALID")

# Card number validation test.


def test_card_number(card_num):
    if card_num < 10**17 and card_num >= 10**14:
        evenDigitsSum = 0
        evenDigits = 0
        oddDigits = 0
        oddDigitsSum = 0
        i = 100
        while i < 10**17:
            evenDigits = (((card_num % i) - (card_num % (i / 10))) / (i / 10)) * 2
            evenDigitsSum = evenDigitsSum + (evenDigits % 10 + ((evenDigits % 100) - (evenDigits % 10)) / 10)
            oddDigits = ((card_num % (i / 10)) - (card_num % (i / 100))) / (i / 100)
            oddDigitsSum = oddDigitsSum + oddDigits
            i *= 100
        if (evenDigitsSum + oddDigitsSum) % 10 == 0:
            return True
        else:
            return False
    else:
        return False


main()